src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+            // normalize to null for NO RULE
+            String ruleName = legend.getRule(); // was null
+
+            gt2Style = resizeForDPI(request, gt2Style);
+
+            final FeatureTypeStyle[] ftStyles =
+                    gt2Style.featureTypeStyles().toArray(new FeatureTypeStyle[0]);
+            final double scaleDenominator = request.getScale();
+
+            Rule[] applicableRules;
+            if (ruleName != null) {
+                Rule rule = LegendUtils.getRule(ftStyles, ruleName);
+                if (rule == null) {
+                    throw new ServiceException(
+                            "Specified style does not contains a rule named " + ruleName);
+                }
+                applicableRules = new Rule[] {rule};
+            } else {
+                applicableRules = LegendUtils.getApplicableRules(ftStyles, scaleDenominator);
+            }
+
+            ArrayList<JSONObject> jRules = new ArrayList<>();
+            for (Rule rule : applicableRules) {
+                JSONObject jRule = new JSONObject();
+                String name = rule.getName();
+                if (name != null && !name.isEmpty()) {
+                    jRule.element(NAME, name);
+                }
+                InternationalString title = rule.getDescription().getTitle();
+                if (title != null) {
+                    jRule.element(TITLE, title.toString(locale));
+                }
+                InternationalString abs = rule.getDescription().getAbstract();
+                if (abs != null) {
+                    jRule.element(ABSTRACT, abs.toString(locale));
+                }
+                Filter filter = rule.getFilter();
+                if (filter != null) {
+                    jRule.element(FILTER, "[" + CQL.toCQL(filter) + "]");
+                }
+                if (rule.isElseFilter()) {
+                    jRule.element(ELSE_FILTER, "true");
+                }
+                JSONArray jSymbolizers = new JSONArray();
+                if (layer != null) {
+                    feature = getSampleFeatureForRule(layer, null, rule);
+                } else {
+                    feature = null;
+                }
+                List<Symbolizer> symbolizers2 = rule.symbolizers();
+                for (Symbolizer symbolizer : symbolizers2) {
+                    JSONObject jSymb = new JSONObject();
+                    JSONObject symb = processSymbolizer(symbolizer);
+                    jSymb.element(symbolizerNames.get(symbolizer.getClass()), symb);
+                    jSymbolizers.add(jSymb);
+                }
+                org.opengis.style.GraphicLegend l = rule.getLegend();
+                if (l != null) {
+                    for (GraphicalSymbol g : l.graphicalSymbols()) {
+                        JSONObject jGraphicSymb = processGraphicalSymbol(g);
+                        String href =
+                                IconPropertyExtractor.extractProperties(
```

Comment:
Does this actually work? Tested?

Reply:
yes, actually returns an image if you try it with a real style